sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs

```diff
@@ -0,0 +1,449 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+
+namespace Azure.Storage.Blobs
+{
+    internal class PartitionedUploader
+    {
+        // The client we use to do the actual uploading
+        private readonly BlockBlobClient _client;
+
+        // The maximum number of simultaneous workers
+        private readonly int _maxWorkerCount;
+
+        // A pool of memory we use to partition the stream into blocks
+        private readonly ArrayPool<byte> _arrayPool;
+
+        // The size we use to determine whether to upload as a single PUT BLOB
+        // request or stage as multiple blocks.
+        private readonly long _singleBlockThreshold;
```

Comment:
Please rename this field. The current name makes me confuse it's purpose with `_blockSize`.

Reply:
That's the name you used to use for this: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs#L969

I think the name is clear, but I can change it if you have any suggestions.

Reply:
Okay, I'll rename it.